blas/gonum/gemv.go

```diff
@@ -0,0 +1,184 @@
+		lenX = n
+		lenY = m
+	}
+	if (incX > 0 && (lenX-1)*incX >= len(x)) || (incX < 0 && (1-lenX)*incX >= len(x)) {
+		panic(badX)
+	}
+	if (incY > 0 && (lenY-1)*incY >= len(y)) || (incY < 0 && (1-lenY)*incY >= len(y)) {
+		panic(badY)
+	}
+	if lda*(m-1)+n > len(a) || lda < max(1, n) {
+		panic(badLdA)
+	}
+
+	// Quick return if possible
+	if m == 0 || n == 0 || (alpha == 0 && beta == 1) {
+		return
+	}
+
+	if alpha == 0 {
+		// First form y = beta * y
+		if incY > 0 {
+			Implementation{}.Dscal(lenY, beta, y, incY)
+		} else {
+			Implementation{}.Dscal(lenY, beta, y, -incY)
+		}
+		return
+	}
+
+	// Form y = alpha * A * x + y
+	if tA == blas.NoTrans {
+		f64.GemvN(uintptr(m), uintptr(n), alpha, a, uintptr(lda), x, uintptr(incX), beta, y, uintptr(incY))
+		return
+	}
+	// Cases where a is transposed.
+	f64.GemvT(uintptr(m), uintptr(n), alpha, a, uintptr(lda), x, uintptr(incX), beta, y, uintptr(incY))
+}
+
+// Sgemv computes
+//  y = alpha * A * x + beta * y    if tA = blas.NoTrans
+//  y = alpha * A^T * x + beta * y  if tA = blas.Trans or blas.ConjTrans
+// where A is an m×n dense matrix, x and y are vectors, and alpha and beta are scalars.
+//
+// Float32 implementations are autogenerated and not directly tested.
+func (Implementation) Sgemv(tA blas.Transpose, m, n int, alpha float32, a []float32, lda int, x []float32, incX int, beta float32, y []float32, incY int) {
+	if tA != blas.NoTrans && tA != blas.Trans && tA != blas.ConjTrans {
+		panic(badTranspose)
+	}
+	if m < 0 {
+		panic(mLT0)
+	}
+	if n < 0 {
+		panic(nLT0)
+	}
+	if lda < max(1, n) {
+		panic(badLdA)
+	}
+
+	if incX == 0 {
+		panic(zeroIncX)
+	}
+	if incY == 0 {
+		panic(zeroIncY)
+	}
+	// Set up indexes
+	lenX := m
+	lenY := n
+	if tA == blas.NoTrans {
+		lenX = n
+		lenY = m
+	}
+	if (incX > 0 && (lenX-1)*incX >= len(x)) || (incX < 0 && (1-lenX)*incX >= len(x)) {
+		panic(badX)
+	}
+	if (incY > 0 && (lenY-1)*incY >= len(y)) || (incY < 0 && (1-lenY)*incY >= len(y)) {
+		panic(badY)
+	}
+	if lda*(m-1)+n > len(a) || lda < max(1, n) {
+		panic(badLdA)
+	}
+
+	// Quick return if possible
+	if m == 0 || n == 0 || (alpha == 0 && beta == 1) {
+		return
+	}
+
+	var kx, ky int
+	if incX > 0 {
+		kx = 0
+	} else {
+		kx = -(lenX - 1) * incX
+	}
+	if incY > 0 {
+		ky = 0
+	} else {
+		ky = -(lenY - 1) * incY
+	}
+
+	// First form y = beta * y
+	if incY > 0 {
+		Implementation{}.Sscal(lenY, beta, y, incY)
+	} else {
+		Implementation{}.Sscal(lenY, beta, y, -incY)
+	}
+
+	if alpha == 0 {
+		return
+	}
+
+	// Form y = alpha * A * x + y
+	if tA == blas.NoTrans {
+		if incX == 1 && incY == 1 {
+			for i := 0; i < m; i++ {
+				y[i] += alpha * f32.DotUnitary(a[lda*i:lda*i+n], x)
+			}
+			return
+		}
+		iy := ky
+		for i := 0; i < m; i++ {
+			y[iy] += alpha * f32.DotInc(x, a[lda*i:lda*i+n], uintptr(n), uintptr(incX), 1, uintptr(kx), 0)
+			iy += incY
+		}
+		return
+	}
+	// Cases where a is transposed.
+	if incX == 1 && incY == 1 {
+		for i := 0; i < m; i++ {
+			tmp := alpha * x[i]
```

Comment:
just a naive drive-by comment...

I am unsure about this possible premature optimization but, wouldn't it help to have just before the `for`-loop:
```go
_ = x[m]
```
to help with all the bounds checks?

Reply:
Are you interested in providing bench data for that? I think that would be a separate PR though.